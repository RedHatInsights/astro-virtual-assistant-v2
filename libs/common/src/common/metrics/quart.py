import logging
import time
from typing import Callable, Optional

import quart
from quart import Quart, request, g
from quart_schema import hide
from aioprometheus import Registry, render, Counter, Histogram

QUART_EXTENSION_METRIC_REGISTRY = "metric_registry"


def register_app(app: Quart, metrics_path="/metrics"):
    registry = Registry()
    app.extensions[QUART_EXTENSION_METRIC_REGISTRY] = registry

    @app.route(metrics_path)
    @hide
    async def handle_metrics():
        content, http_headers = render(registry, request.headers.getlist("accept"))
        return content, http_headers


def get_registry(app: Quart):
    if QUART_EXTENSION_METRIC_REGISTRY not in app.extensions:
        raise KeyError(
            f"Metrics registry '{QUART_EXTENSION_METRIC_REGISTRY}' is missing. Please ensure it is registered via 'register_app'."
        )
    return app.extensions[QUART_EXTENSION_METRIC_REGISTRY]


def register_http_metrics(
    app: Quart,
    app_name: str,
    accept_paths: Optional[Callable[[quart.Request], bool]] = None,
):
    registry = get_registry(app)
    http_requests_total = Counter(
        "http_requests_total",
        "Total number of HTTP requests",
        const_labels={"app": app_name},
        registry=registry,
    )
    http_request_duration = Histogram(
        "http_request_duration_seconds",
        "HTTP request latency in seconds",
        buckets=[0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1.0, 2.5, 5.0],
        const_labels={"app": app_name},
        registry=registry,
    )

    @app.before_request
    async def before():
        g.metrics_request_start_time = time.monotonic()

    @app.after_request
    async def after(response):
        if accept_paths is None or accept_paths(request):
            duration = time.monotonic() - g.metrics_request_start_time

            labels = {
                "method": request.method,
                "path": request.url_rule.rule
                if request.url_rule is not None
                else "<unknown>",
                "status": str(response.status_code),
            }

            try:
                http_requests_total.inc(labels)
                http_request_duration.observe(labels, duration)
            except Exception as e:
                logging.getLogger(__name__).error(
                    "Failed to send platform_request metrics", e
                )

        return response
