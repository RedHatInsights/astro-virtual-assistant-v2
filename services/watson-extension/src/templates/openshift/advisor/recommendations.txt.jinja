{%- if recommendations -%}
Here are your most recent {{ category }} from OpenShift for Advisor.
{% for recommendation in recommendations %}
 {{ loop.index }}. [{{ recommendation.description }}]({{ recommendation.link }})
{%- endfor %}

You can see additional {{ category }} on the [OpenShift for Advisor dashboard]({{ dashboard_link }}).
{%- else -%}
You don't have any {{ category }} right now.
{%- endif -%}