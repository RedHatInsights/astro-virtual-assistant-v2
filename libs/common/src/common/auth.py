from __future__ import annotations

import binascii

from quart import request, jsonify
import base64
import functools
import json


def check_identity(identity_header):
    try:
        base64.b64decode(identity_header)
    except binascii.Error:
        return False
    return True


def require_identity_header(func):
    @functools.wraps(func)
    async def wrapper(*args, **kwargs):
        identity_header = request.headers.get("x-rh-identity")
        if not identity_header or not check_identity(identity_header):
            return jsonify(message="Invalid x-rh-identity"), 401
        return await func(*args, **kwargs)

    return wrapper


def assistant_user_id(identity):
    decoded_identity = base64.b64decode(identity).decode("utf8")
    identity_wrapper = json.loads(decoded_identity)
    identity_json = identity_wrapper.get("identity")

    org_id = identity_json.get("org_id")
    identity_type = identity_json.get("type")
    if identity_type == "ServiceAccount":
        user_id = identity_json.get("service_account").get("user_id")
    elif identity_type == "User":
        user_id = identity_json.get("user").get("user_id")
    elif identity_type == "System":
        user_id = identity_json.get("system").get("cn")

    return f"{org_id}/{user_id}"
