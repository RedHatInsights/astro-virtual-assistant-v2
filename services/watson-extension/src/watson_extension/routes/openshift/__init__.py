from quart import Blueprint
from . import advisor

blueprint = Blueprint("openshift", __name__, url_prefix="/openshift")

blueprint.register_blueprint(advisor.blueprint)
