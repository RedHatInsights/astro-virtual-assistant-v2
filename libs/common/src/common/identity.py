import abc

import quart
from werkzeug.exceptions import BadRequest

import injector
from common.session_storage import SessionStorage


class AbstractUserIdentityProvider(abc.ABC):
    async def get_user_identity(self) -> str: ...


class QuartWatsonExtensionUserIdentityProvider(AbstractUserIdentityProvider):
    def __init__(
        self, request: quart.Request, session_storage: injector.Inject[SessionStorage]
    ):
        self.request = request
        self.session_storage = session_storage

    async def get_user_identity(self):
        session_header_name = "x-rh-session-id"
        if session_header_name not in self.request.headers:
            raise BadRequest(f"Missing ${session_header_name}")

        session_id = self.request.headers[session_header_name]
        return (await self.session_storage.get(session_id)).user_identity


class QuartRedHatUserIdentityProvider(AbstractUserIdentityProvider):
    def __init__(self, request: quart.Request):
        self.request = request

    async def get_user_identity(self) -> str:
        session_header_name = "x-rh-identity"
        if session_header_name not in self.request.headers:
            raise BadRequest(f"Missing ${session_header_name}")

        return self.request.headers[session_header_name]


class FixedUserIdentityProvider(AbstractUserIdentityProvider):
    async def get_user_identity(self):
        """
        Fixed user identity equivalent to:
        {
           "identity":{
              "account_number":"account123",
              "org_id":"org123",
              "type":"User",
              "user":{
                 "is_org_admin":true,
                 "is_internal": true,
                 "user_id":"1234567890",
                 "username":"astro"
              },
              "internal":{
                 "org_id":"org123"
              }
           }
        }
        """

        return "eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJvcmdfaWQiOiJvcmcxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlLCJpc19pbnRlcm5hbCI6dHJ1ZSwidXNlcl9pZCI6IjEyMzQ1Njc4OTAiLCJ1c2VybmFtZSI6ImFzdHJvIn0sImludGVybmFsIjp7Im9yZ19pZCI6Im9yZzEyMyJ9fX0="
