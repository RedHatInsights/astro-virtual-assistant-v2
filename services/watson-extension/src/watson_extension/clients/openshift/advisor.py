import abc
import injector
from dataclasses import dataclass
from typing import List, Optional

from common.identity import AbstractUserIdentityProvider
from common.platform_request import AbstractPlatformRequest
from watson_extension.clients import AdvisorOpenshiftURL


@dataclass
class Cluster:
    id: str
    name: str
    last_checked_at: Optional[str] = None


@dataclass
class Workload:
    cluster_display_name: str
    cluster_id: str
    namespace_id: str
    last_checked_at: Optional[str] = None


@dataclass
class Recommendation:
    id: str
    description: str
    total_risk: int


class AdvisorClient(abc.ABC):
    @abc.abstractmethod
    async def get_clusters(self) -> List[Cluster]: ...

    @abc.abstractmethod
    async def get_workloads(self) -> List[Workload]: ...

    @abc.abstractmethod
    async def get_recommendations(self) -> List[Recommendation]: ...


class AdvisorClientHttp(AdvisorClient):
    def __init__(
        self,
        advisor_url: injector.Inject[AdvisorOpenshiftURL],
        user_identity_provider: injector.Inject[AbstractUserIdentityProvider],
        platform_request: injector.Inject[AbstractPlatformRequest],
    ):
        super().__init__()
        self.advisor_url = advisor_url
        self.user_identity_provider = user_identity_provider
        self.platform_request = platform_request

    async def get_clusters(self) -> List[Cluster]:
        response = await self.platform_request.get(
            self.advisor_url,
            "/api/insights-results-aggregator/v2/clusters",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        clusters = [
            Cluster(
                id=c["cluster_id"],
                name=c["cluster_name"],
                last_checked_at=c["last_checked_at"]
                if "last_checked_at" in c
                else None,
            )
            for c in content["data"]
        ]

        clusters.sort(
            key=lambda c: [c.last_checked_at] if c.last_checked_at is not None else [],
            reverse=True,
        )

        return clusters[:3]

    async def get_workloads(self) -> List[Workload]:
        response = await self.platform_request.get(
            self.advisor_url,
            "/api/insights-results-aggregator/v2/namespaces/dvo",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()
        if content["status"] != "ok":
            raise RuntimeError(
                f"Received invalid status from openshift_advisor/workloads: {content['status']}"
            )

        workloads = [
            Workload(
                cluster_id=w["cluster"]["uuid"],
                cluster_display_name=w["cluster"]["display_name"],
                namespace_id=w["namespace"]["uuid"],
                last_checked_at=w.get("metadata", {}).get("last_checked_at", None),
            )
            for w in content["workloads"]
        ]

        workloads.sort(
            key=lambda w: [w.last_checked_at] if w.last_checked_at is not None else [],
            reverse=True,
        )

        return workloads[:3]

    async def get_recommendations(self) -> List[Recommendation]:
        response = await self.platform_request.get(
            self.advisor_url,
            "/api/insights-results-aggregator/v2/rule?impacting=true",
            user_identity=await self.user_identity_provider.get_user_identity(),
        )
        response.raise_for_status()

        content = await response.json()

        rules = [
            Recommendation(
                id=r["rule_id"],
                description=r["description"],
                total_risk=r["total_risk"],
            )
            for r in content["recommendations"]
        ]
        rules.sort(key=lambda r: r.total_risk, reverse=True)

        return rules[:3]
