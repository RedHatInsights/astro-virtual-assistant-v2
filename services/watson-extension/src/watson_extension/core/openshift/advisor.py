import enum
from dataclasses import dataclass
from typing import List

import injector

from watson_extension.clients.openshift.advisor import AdvisorClient


class RecommendationCategory(enum.Enum):
    """
    Categories to group different searching queries for rhel advisor recommendations
    """

    RECOMMENDATION = "recommendation"
    CLUSTER = "cluster"
    WORKLOAD = "workload"


@dataclass
class Recommendation:
    description: str
    link: str


@dataclass
class AdvisorRecommendationResponse:
    category: RecommendationCategory
    dashboard_link: str
    recommendations: List[Recommendation]


class AdvisorCore:
    def __init__(self, advisor_client: injector.Inject[AdvisorClient]):
        self.advisor_client = advisor_client

    async def get_recommendations(
        self, category_type: RecommendationCategory
    ) -> AdvisorRecommendationResponse:
        if category_type == RecommendationCategory.RECOMMENDATION:
            recommendations = await self.advisor_client.get_recommendations()
            return AdvisorRecommendationResponse(
                category=category_type,
                dashboard_link="/openshift/insights/advisor/recommendations",
                recommendations=[
                    Recommendation(
                        description=r.description,
                        link=f"/openshift/insights/advisor/recommendations/{r.id}",
                    )
                    for r in recommendations
                ],
            )
        elif category_type == RecommendationCategory.CLUSTER:
            clusters = await self.advisor_client.get_clusters()
            return AdvisorRecommendationResponse(
                category=category_type,
                dashboard_link="/openshift/insights/advisor/clusters",
                recommendations=[
                    Recommendation(
                        description=c.name,
                        link=f"/openshift/insights/advisor/clusters/{c.id}",
                    )
                    for c in clusters
                ],
            )
        elif category_type == RecommendationCategory.WORKLOAD:
            workloads = await self.advisor_client.get_workloads()
            return AdvisorRecommendationResponse(
                category=category_type,
                dashboard_link="/openshift/insights/advisor/workloads",
                recommendations=[
                    Recommendation(
                        description=w.cluster_display_name,
                        link=f"/openshift/insights/advisor/workloads/{w.cluster_id}/{w.namespace_id}",
                    )
                    for w in workloads
                ],
            )
        else:
            raise RuntimeError(
                f"Unexpected openshift advisor recommendation type: {category_type}"
            )
