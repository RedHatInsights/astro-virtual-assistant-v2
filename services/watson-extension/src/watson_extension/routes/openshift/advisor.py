import injector
from pydantic import BaseModel

from quart import Blueprint, render_template
from quart_schema import validate_response, validate_querystring, document_headers

from watson_extension.core.openshift.advisor import (
    RecommendationCategory,
    AdvisorCore,
)
from watson_extension.routes import RHSessionIdHeader

blueprint = Blueprint("advisor", __name__, url_prefix="/advisor")


class RecommendationsRequestQuery(BaseModel):
    category: RecommendationCategory


class RecommendationsResponse(BaseModel):
    response: str


@blueprint.get("/recommendations")
@validate_querystring(RecommendationsRequestQuery)
@validate_response(RecommendationsResponse)
@document_headers(RHSessionIdHeader)
async def recommendations(
    query_args: RecommendationsRequestQuery,
    advisor_service: injector.Inject[AdvisorCore],
) -> RecommendationsResponse:
    recommendation_response = await advisor_service.get_recommendations(
        query_args.category
    )

    if recommendation_response.category == RecommendationCategory.RECOMMENDATION:
        category_name = "recommendations"
    elif recommendation_response.category == RecommendationCategory.WORKLOAD:
        category_name = "workloads"
    elif recommendation_response.category == RecommendationCategory.CLUSTER:
        category_name = "clusters"
    else:
        raise ValueError(f"Invalid category: {recommendation_response.category}")

    return RecommendationsResponse(
        response=await render_template(
            "openshift/advisor/recommendations.txt.jinja",
            recommendations=recommendation_response.recommendations,
            dashboard_link=recommendation_response.dashboard_link,
            category=category_name,
        )
    )
