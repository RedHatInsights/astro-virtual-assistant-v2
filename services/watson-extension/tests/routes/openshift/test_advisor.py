from unittest.mock import MagicMock

import injector
import pytest
from quart.typing import TestClientProtocol

from watson_extension.clients.openshift.advisor import (
    AdvisorClient,
    Recommendation,
    Workload,
    Cluster,
)
from ..common import app_with_blueprint

from watson_extension.routes.openshift.advisor import blueprint
from ... import async_value, get_test_template


@pytest.fixture
async def advisor_client() -> MagicMock:
    return MagicMock(AdvisorClient)


@pytest.fixture
async def test_client(advisor_client) -> TestClientProtocol:
    def injector_binder(binder: injector.Binder):
        binder.bind(AdvisorClient, advisor_client)

    return app_with_blueprint(blueprint, injector_binder).test_client()


async def test_recommendations(test_client, advisor_client) -> None:
    advisor_client.get_recommendations = MagicMock(
        return_value=async_value(
            [
                Recommendation(
                    description="recommendation 1", total_risk=13, id="my-id"
                ),
                Recommendation(
                    description="recommendation 2", total_risk=5, id="my-id-2"
                ),
            ]
        )
    )

    response = await test_client.get(
        "/advisor/recommendations", query_string={"category": "recommendation"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template(
        "openshift/advisor/recommendations.txt"
    )


async def test_workloads(test_client, advisor_client) -> None:
    advisor_client.get_workloads = MagicMock(
        return_value=async_value(
            [
                Workload(
                    cluster_id="cluster-001",
                    cluster_display_name="My cluster1",
                    namespace_id="my-namespace",
                    last_checked_at="2025-03-31T13:51:18+00:00",
                ),
                Workload(
                    cluster_id="cluster-001",
                    cluster_display_name="My cluster5",
                    namespace_id="somewhere",
                    last_checked_at="2025-03-30T13:51:18+00:00",
                ),
                Workload(
                    cluster_id="cluster-003",
                    cluster_display_name="My cluster10",
                    namespace_id="etc",
                    last_checked_at="2025-03-29T13:51:18+00:00",
                ),
            ]
        )
    )

    response = await test_client.get(
        "/advisor/recommendations", query_string={"category": "workload"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("openshift/advisor/workload.txt")


async def test_clusters(test_client, advisor_client) -> None:
    advisor_client.get_clusters = MagicMock(
        return_value=async_value(
            [
                Cluster(
                    id="007", name="Bond", last_checked_at="2025-03-31T13:51:18+00:00"
                ),
                Cluster(
                    id="one", name="Johnny", last_checked_at="2025-03-30T13:51:18+00:00"
                ),
            ]
        )
    )

    response = await test_client.get(
        "/advisor/recommendations", query_string={"category": "cluster"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template("openshift/advisor/cluster.txt")


async def test_recommendations_none(test_client, advisor_client) -> None:
    advisor_client.get_recommendations = MagicMock(return_value=async_value([]))

    response = await test_client.get(
        "/advisor/recommendations", query_string={"category": "recommendation"}
    )
    assert response.status == "200 OK"
    data = await response.get_json()
    assert data["response"] == get_test_template(
        "openshift/advisor/no-recommendations.txt"
    )
