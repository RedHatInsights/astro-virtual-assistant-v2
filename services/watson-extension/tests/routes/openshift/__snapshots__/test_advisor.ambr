# serializer version: 1
# name: test_clusters
  '''
  Here are your most recent clusters from OpenShift for Advisor.
  
   1. [Bond](/openshift/insights/advisor/clusters/007)
   2. [Johnny](/openshift/insights/advisor/clusters/one)
  
  You can see additional clusters on the [OpenShift for Advisor dashboard](/openshift/insights/advisor/clusters).
  '''
# ---
# name: test_recommendations
  '''
  Here are your most recent recommendations from OpenShift for Advisor.
  
   1. [recommendation 1](/openshift/insights/advisor/recommendations/my-id)
   2. [recommendation 2](/openshift/insights/advisor/recommendations/my-id-2)
  
  You can see additional recommendations on the [OpenShift for Advisor dashboard](/openshift/insights/advisor/recommendations).
  '''
# ---
# name: test_recommendations_none
  "You don't have any recommendations right now."
# ---
# name: test_workloads
  '''
  Here are your most recent workloads from OpenShift for Advisor.
  
   1. [My cluster1](/openshift/insights/advisor/workloads/cluster-001/my-namespace)
   2. [My cluster5](/openshift/insights/advisor/workloads/cluster-001/somewhere)
   3. [My cluster10](/openshift/insights/advisor/workloads/cluster-003/etc)
  
  You can see additional workloads on the [OpenShift for Advisor dashboard](/openshift/insights/advisor/workloads).
  '''
# ---
